### PR TITLE
Shrestha/add op sparse softmax cross entropy with logits

### PR DESCRIPTION
**Pull request by shresthamalik**
Implement op SparseSoftmaxCrossEntropyWithLogits
+ Some clean up on ReluGrad python test case
+ Arranged the registeration of ops in alphabetical order
+ some changes(noise) added due to clang format

### CODE REVIEW

File: src/ngraph_builder.cc [hunk 1]
```diff
@@ -1864,6 +1867,59 @@ static tf::Status TranslateSoftmaxOp(const tf::Node* op,
   return tf::Status::OK();
 }
 
+static tf::Status TranslateSparseSoftmaxCrossEntropyWithLogitsOp(
+    const tf::Node* op, Builder::OpMap& ng_op_map) {
+  // TF op Inputs:
```
Comment (avijit-nervana) [inline src/ngraph_builder.cc]: This is very good. This is how the op implementation should be documented.

File: test/python/test_sparse_softmax_cross_entropy_with_logits.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import numpy as np
+from common import NgraphTest
+
+
+class TestSparseSoftmaxCrossEntropyWithLogitsOperations(NgraphTest):
+  def test_sparse_softmax_cross_entropy_with_logits_2d(self):
+    num_classes = 10
+    batch_size = 1000
+    total_size = num_classes * batch_size
+    labels = constant_op.constant(self.generate_random_numbers(batch_size, 0, num_classes-1, "DTYPE_INT"), shape = [batch_size])
```
Comment (avijit-nervana) [inline test/python/test_sparse_softmax_cross_entropy_with_logits.py]: Is it possible to break the long lines into the ones with 80 col width?

File: test/tf_exec.cpp [hunk 3]
```diff
@@ -108,13 +108,11 @@ TEST(tf_exec, axpy) {
 }
 #endif
 
-void AssertTensorEquals(tf::Tensor T1, tf::Tensor T2) {
+void AssertTensorEquals(tf::Tensor& T1, tf::Tensor& T2) {
```
Comment (avijit-nervana) [inline test/tf_exec.cpp]: Good catch!

File: src/ngraph_builder.cc [hunk 4]
```diff
@@ -1864,6 +1867,59 @@ static tf::Status TranslateSoftmaxOp(const tf::Node* op,
...
+
+  ng_axes.insert(1);
+  auto ng_features_softmax = make_shared<ng::op::Softmax>(ng_features, ng_axes);
+
+  // Compute one-hot encoding of labels : y_true
+  // one_hot_axis is 1 as we are restricting the shape to be 2D [BatchSize,
+  // NumClasses]
+  auto ng_onehot_labels =
+      make_shared<ng::op::OneHot>(ng_labels, ng_features_shape, 1);
+  auto ng_onehot_labels_float = make_shared<ng::op::Convert>(
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply src/ngraph_builder.cc]: Sure. The expected type of labels is int32/int64 and of features is float32/float64, so the convert would be required in most cases. But no harm adding the check.

File: test/python/test_sparse_softmax_cross_entropy_with_logits.py (hunk 2 above)
Comment (shresthamalik) [reply test/python/test_sparse_softmax_cross_entropy_with_logits.py]: Yes, i will work on this.

File: src/ngraph_builder.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (aprocter) [reply src/ngraph_builder.cc]: Oh, okay, in that case I'd agree there's no need for the check. Up to you.

**Comment by shresthamalik** [comment]
@aprocter @avijit-nervana  I have made the changes as discussed. Please review

File: src/ngraph_builder.cc [hunk 5]
```diff
@@ -1921,6 +1921,7 @@ static tf::Status TranslateSoftmaxOp(const tf::Node* op,
   return tf::Status::OK();
 }
 
+/*
```
Comment (aprocter) [inline src/ngraph_builder.cc]: Let's delete the commented-out code before the PR goes in.

File: test/python/test_sparse_softmax_cross_entropy_with_logits.py [hunk 6]
```diff
@@ -34,8 +34,10 @@ def test_sparse_softmax_cross_entropy_with_logits_2d(self):
     num_classes = 10
     batch_size = 1000
     total_size = num_classes * batch_size
-    labels = constant_op.constant(self.generate_random_numbers(batch_size, 0, num_classes-1, "DTYPE_INT"), shape = [batch_size])
-    features = constant_op.constant(self.generate_random_numbers(total_size, 0.0, 1.0), shape =[batch_size, num_classes])
+    labels = constant_op.constant(self.generate_random_numbers(batch_size, 0,
+    num_classes-1, "DTYPE_INT"), shape = [batch_size])
```
Comment (aprocter) [inline test/python/test_sparse_softmax_cross_entropy_with_logits.py]: The indentation looks weird here (maybe GitHub is making a mess of it...)

File: test/python/test_sparse_softmax_cross_entropy_with_logits.py [hunk 7]
```diff
@@ -34,8 +34,10 @@ def test_sparse_softmax_cross_entropy_with_logits_2d(self):
...
+    features = constant_op.constant(self.generate_random_numbers(total_size,
+    0.0, 1.0), shape =[batch_size, num_classes])
```
Comment (aprocter) [inline test/python/test_sparse_softmax_cross_entropy_with_logits.py]: Same comment on indentation here

File: src/ngraph_builder.cc [hunk 8]
```diff
@@ -1926,51 +1926,104 @@ static tf::Status TranslateSparseSoftmaxCrossEntropyWithLogitsOp(
...
+        " Logits/Features must be shape 2-D, but got shape ",
+        ng::join(ng_features_shape), " while building op ", op->type_string());
+  }
 
-  if (shape_size != 2) {
+  // Labels must be 1-d shape
+  if (ng_labels_shape.size() != 1) {
     return tf::errors::InvalidArgument(
-        "TF Softmax logits must be 2-dimensional");
+        " Labels must be shape 1-D, but got shape ", ng::join(ng_labels_shape),
```
Comment (avijit-nervana) [inline src/ngraph_builder.cc]: Great error messages! Very descriptive.
